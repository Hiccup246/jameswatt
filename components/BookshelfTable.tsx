import React, { ReactElement, useEffect, useState } from "react";
import Image from "next/image";
import { ThemeContext } from "./ThemeProvider";
import { BookStatus } from "../constants/Books";

const whiteClipboardChecked = (
  <div className="mx-auto h-5 w-5">
    <Image
      src={"/clipboard-checked-white.webp"}
      width={20}
      height={20}
      sizes="20px"
      alt="Completed Book Icon"
    />
  </div>
);

const blackClipboardChecked = (
  <div className="mx-auto h-5 w-5">
    <Image
      src={"/clipboard-checked-black.webp"}
      width={20}
      height={20}
      sizes="20px"
      alt="Completed Book Icon"
    />
  </div>
);

const whiteAnimatedBook = (
  <div className="mx-auto h-5 w-5">
    <Image
      unoptimized
      src={"/animated-book-white.gif"}
      width={20}
      height={20}
      sizes="20px"
      alt="Currently Reading Icon"
    />
  </div>
);

const blackAnimatedBook = (
  <div className="mx-auto h-5 w-5">
    <Image
      unoptimized
      src={"/animated-book-black.gif"}
      width={20}
      height={20}
      sizes="20px"
      alt="Currently Reading Icon"
    />
  </div>
);

export default function BookShelfTable({
  books = [],
  shortViewBooks = 0,
  shortView = false,
}: {
  books?: Book[];
  shortViewBooks?: number;
  shortView?: boolean;
}) {
  const [clipboardCheckedIcon, setClipboardCheckedIcon] =
    useState<ReactElement>();
  const [animatedBook, setAnimatedBook] = useState<ReactElement>();
  const previewBooks = shortViewBooks < 0 ? 0 : shortViewBooks;
  const { siteTheme } = React.useContext(ThemeContext);

  useEffect(() => {
    if (siteTheme == "dark") {
      setAnimatedBook(whiteAnimatedBook);
      setClipboardCheckedIcon(whiteClipboardChecked);
    } else {
      setAnimatedBook(blackAnimatedBook);
      setClipboardCheckedIcon(blackClipboardChecked);
    }
  }, [siteTheme]);

  return (
    <table className="w-full">
      <thead className="bg-brown dark:bg-darkgrey sticky top-0">
        <tr>
          <th className="pb-2.5 text-left text-sm">Book</th>
          <th className="max-smaller:hidden w-24 pb-2.5 text-left text-sm">
            Genre
          </th>
          <th className="pb-2.5 text-center text-sm">Status</th>
        </tr>
      </thead>
      <tbody className="tr-gap">
        {(!!shortView ? books.slice(0, previewBooks) : books).map(
          (book: Book, index: number) => {
            return (
              <tr key={book.name + index}>
                <td className="flex flex-col justify-start pr-2.5 align-top break-keep">
                  <h2 className="my-0 max-w-lg font-bold">{book.name}</h2>
                  <div className="flex flex-col">
                    <div className="text-sm">{book.author}</div>
                    <div
                      title={
                        book.status == BookStatus.Read
                          ? "Date Completed"
                          : "Date Started"
                      }
                      className="max-smaller:order-first order-last text-sm"
                    >
                      {book.dateCompleted || book.dateStarted}
                    </div>
                  </div>
                </td>

                <td className="max-smaller:inline align-top text-sm">
                  {book.genre}
                </td>

                <td className="align-top">
                  {book.status == BookStatus.Read
                    ? clipboardCheckedIcon
                    : animatedBook}
                </td>
              </tr>
            );
          },
        )}
      </tbody>
    </table>
  );
}
