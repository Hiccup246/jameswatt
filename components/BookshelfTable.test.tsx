import BookshelfTable from "./BookshelfTable";
import { render, screen } from "@testing-library/react";

const headerRows = 1;
const mockBooks: Book[] = [
  {
    name: "The Black Swan: The Impact of the Highly Improbable",
    author: "Nassim Nicholas Taleb",
    genre: "Nonfiction",
    status: "Reading",
    dateStarted: "26/09/2022",
    dateCompleted: "",
  },
  {
    name: "The Pragmatic Programmer: From Journeyman to Master",
    author: "Andy Hunt, Dave Thomas",
    genre: "Programming",
    status: "Reading",
    dateStarted: "",
    dateCompleted: "",
  },
  {
    name: "Wordslut: A Feminist Guide to Taking Back the English Language",
    author: "Amanda Montell",
    genre: "Nonfiction",
    status: "Read",
    dateStarted: "12/09/2022",
    dateCompleted: "25/09/2022",
  },
  {
    name: "Steve Jobs",
    author: "Walter Isaacson",
    genre: "Biography",
    status: "Read",
    dateStarted: "24/07/2022",
    dateCompleted: "11/09/2022",
  },
  {
    name: "Brave New World",
    author: "Aldous Huxley",
    genre: "Fiction",
    status: "Read",
    dateStarted: "",
    dateCompleted: "",
  },
  {
    name: "Modern Conspiracies in America: Separating Fact from Fiction",
    author: "Michael D. Gambone",
    genre: "Nonfiction",
    status: "Read",
    dateStarted: "",
    dateCompleted: "26/06/2022",
  },
];

describe("BookshelfSection", () => {
  Object.defineProperty(document, "documentElement", {
    configurable: true,
    value: document.createElement("document"),
  });

  it("correctly renders all books", () => {
    const { container } = render(<BookshelfTable books={mockBooks} />);

    expect(container).toMatchSnapshot();
  });

  it("renders the correct table headers", () => {
    render(<BookshelfTable />);

    const bookHeader = screen.getByText("Book");
    const genreHeader = screen.getByText("Genre");
    const statusHeader = screen.getByText("Status");

    expect(bookHeader).toBeInTheDocument();
    expect(genreHeader).toBeInTheDocument();
    expect(statusHeader).toBeInTheDocument();
  });

  it("renders no books when none are given", () => {
    render(<BookshelfTable />);

    expect(screen.getAllByRole("row").length - headerRows).toBe(0);
  });

  it("renders shortViewBooks when shortView is true", () => {
    render(<BookshelfTable books={mockBooks} shortViewBooks={4} shortView />);

    expect(screen.getAllByRole("row").length - headerRows).toBe(4);
  });

  it("renders all books when shortViewBooks is set but shortView is false", () => {
    render(<BookshelfTable books={mockBooks} shortViewBooks={4} />);

    expect(screen.getAllByRole("row").length - headerRows).toBe(
      mockBooks.length
    );
  });

  it("renders no books when shortViewBooks is negative and shortView is true", () => {
    render(<BookshelfTable books={mockBooks} shortViewBooks={-1} shortView />);

    expect(screen.getAllByRole("row").length - headerRows).toBe(0);
  });
});
