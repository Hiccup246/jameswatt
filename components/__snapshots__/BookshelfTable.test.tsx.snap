// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookshelfSection correctly renders all books 1`] = `
<table
  className="w-full"
>
  <thead
    className="sticky top-0 bg-brown dark:bg-darkgrey"
  >
    <tr>
      <th
        className="pb-2.5 text-left text-sm"
      >
        Book
      </th>
      <th
        className="w-24 pb-2.5 text-left text-sm max-smaller:hidden"
      >
        Genre
      </th>
      <th
        className="pb-2.5 text-center text-sm"
      >
        Status
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      className="[&>td]:pt-10 [&>td]:first:pt-0"
    >
      <td
        className="flex flex-col justify-start break-keep pr-2.5 align-top"
      >
        <h2
          className="my-0 max-w-lg font-bold"
        >
          The Black Swan: The Impact of the Highly Improbable
        </h2>
        <div
          className="text-sm"
        >
          Nassim Nicholas Taleb
        </div>
      </td>
      <td
        className="align-top text-sm max-smaller:inline"
      >
        Nonfiction
      </td>
      <td
        className="align-top"
      >
        <div
          className="mx-auto h-5 w-5"
        >
          <img
            alt="Currently Reading Icon"
            data-nimg="1"
            decoding="async"
            height={20}
            loading="lazy"
            onError={[Function]}
            onLoad={[Function]}
            sizes="20px"
            src="/_next/image?url=%2Fimg.jpg&w=3840&q=75"
            srcSet="/_next/image?url=%2Fimg.jpg&w=16&q=75 16w, /_next/image?url=%2Fimg.jpg&w=32&q=75 32w, /_next/image?url=%2Fimg.jpg&w=48&q=75 48w, /_next/image?url=%2Fimg.jpg&w=64&q=75 64w, /_next/image?url=%2Fimg.jpg&w=96&q=75 96w, /_next/image?url=%2Fimg.jpg&w=128&q=75 128w, /_next/image?url=%2Fimg.jpg&w=256&q=75 256w, /_next/image?url=%2Fimg.jpg&w=384&q=75 384w, /_next/image?url=%2Fimg.jpg&w=640&q=75 640w, /_next/image?url=%2Fimg.jpg&w=750&q=75 750w, /_next/image?url=%2Fimg.jpg&w=828&q=75 828w, /_next/image?url=%2Fimg.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimg.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimg.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimg.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimg.jpg&w=3840&q=75 3840w"
            style={
              {
                "color": "transparent",
              }
            }
            width={20}
          />
        </div>
      </td>
    </tr>
    <tr
      className="[&>td]:pt-10 [&>td]:first:pt-0"
    >
      <td
        className="flex flex-col justify-start break-keep pr-2.5 align-top"
      >
        <h2
          className="my-0 max-w-lg font-bold"
        >
          The Pragmatic Programmer: From Journeyman to Master
        </h2>
        <div
          className="text-sm"
        >
          Andy Hunt, Dave Thomas
        </div>
      </td>
      <td
        className="align-top text-sm max-smaller:inline"
      >
        Programming
      </td>
      <td
        className="align-top"
      >
        <div
          className="mx-auto h-5 w-5"
        >
          <img
            alt="Currently Reading Icon"
            data-nimg="1"
            decoding="async"
            height={20}
            loading="lazy"
            onError={[Function]}
            onLoad={[Function]}
            sizes="20px"
            src="/_next/image?url=%2Fimg.jpg&w=3840&q=75"
            srcSet="/_next/image?url=%2Fimg.jpg&w=16&q=75 16w, /_next/image?url=%2Fimg.jpg&w=32&q=75 32w, /_next/image?url=%2Fimg.jpg&w=48&q=75 48w, /_next/image?url=%2Fimg.jpg&w=64&q=75 64w, /_next/image?url=%2Fimg.jpg&w=96&q=75 96w, /_next/image?url=%2Fimg.jpg&w=128&q=75 128w, /_next/image?url=%2Fimg.jpg&w=256&q=75 256w, /_next/image?url=%2Fimg.jpg&w=384&q=75 384w, /_next/image?url=%2Fimg.jpg&w=640&q=75 640w, /_next/image?url=%2Fimg.jpg&w=750&q=75 750w, /_next/image?url=%2Fimg.jpg&w=828&q=75 828w, /_next/image?url=%2Fimg.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimg.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimg.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimg.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimg.jpg&w=3840&q=75 3840w"
            style={
              {
                "color": "transparent",
              }
            }
            width={20}
          />
        </div>
      </td>
    </tr>
    <tr
      className="[&>td]:pt-10 [&>td]:first:pt-0"
    >
      <td
        className="flex flex-col justify-start break-keep pr-2.5 align-top"
      >
        <h2
          className="my-0 max-w-lg font-bold"
        >
          Wordslut: A Feminist Guide to Taking Back the English Language
        </h2>
        <div
          className="text-sm"
        >
          Amanda Montell
        </div>
      </td>
      <td
        className="align-top text-sm max-smaller:inline"
      >
        Nonfiction
      </td>
      <td
        className="align-top"
      >
        <div
          className="mx-auto h-5 w-5"
        >
          <img
            alt="Completed Book Icon"
            data-nimg="1"
            decoding="async"
            height={20}
            loading="lazy"
            onError={[Function]}
            onLoad={[Function]}
            sizes="20px"
            src="/_next/image?url=%2Fimg.jpg&w=3840&q=75"
            srcSet="/_next/image?url=%2Fimg.jpg&w=16&q=75 16w, /_next/image?url=%2Fimg.jpg&w=32&q=75 32w, /_next/image?url=%2Fimg.jpg&w=48&q=75 48w, /_next/image?url=%2Fimg.jpg&w=64&q=75 64w, /_next/image?url=%2Fimg.jpg&w=96&q=75 96w, /_next/image?url=%2Fimg.jpg&w=128&q=75 128w, /_next/image?url=%2Fimg.jpg&w=256&q=75 256w, /_next/image?url=%2Fimg.jpg&w=384&q=75 384w, /_next/image?url=%2Fimg.jpg&w=640&q=75 640w, /_next/image?url=%2Fimg.jpg&w=750&q=75 750w, /_next/image?url=%2Fimg.jpg&w=828&q=75 828w, /_next/image?url=%2Fimg.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimg.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimg.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimg.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimg.jpg&w=3840&q=75 3840w"
            style={
              {
                "color": "transparent",
              }
            }
            width={20}
          />
        </div>
      </td>
    </tr>
    <tr
      className="[&>td]:pt-10 [&>td]:first:pt-0"
    >
      <td
        className="flex flex-col justify-start break-keep pr-2.5 align-top"
      >
        <h2
          className="my-0 max-w-lg font-bold"
        >
          Steve Jobs
        </h2>
        <div
          className="text-sm"
        >
          Walter Isaacson
        </div>
      </td>
      <td
        className="align-top text-sm max-smaller:inline"
      >
        Biography
      </td>
      <td
        className="align-top"
      >
        <div
          className="mx-auto h-5 w-5"
        >
          <img
            alt="Completed Book Icon"
            data-nimg="1"
            decoding="async"
            height={20}
            loading="lazy"
            onError={[Function]}
            onLoad={[Function]}
            sizes="20px"
            src="/_next/image?url=%2Fimg.jpg&w=3840&q=75"
            srcSet="/_next/image?url=%2Fimg.jpg&w=16&q=75 16w, /_next/image?url=%2Fimg.jpg&w=32&q=75 32w, /_next/image?url=%2Fimg.jpg&w=48&q=75 48w, /_next/image?url=%2Fimg.jpg&w=64&q=75 64w, /_next/image?url=%2Fimg.jpg&w=96&q=75 96w, /_next/image?url=%2Fimg.jpg&w=128&q=75 128w, /_next/image?url=%2Fimg.jpg&w=256&q=75 256w, /_next/image?url=%2Fimg.jpg&w=384&q=75 384w, /_next/image?url=%2Fimg.jpg&w=640&q=75 640w, /_next/image?url=%2Fimg.jpg&w=750&q=75 750w, /_next/image?url=%2Fimg.jpg&w=828&q=75 828w, /_next/image?url=%2Fimg.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimg.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimg.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimg.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimg.jpg&w=3840&q=75 3840w"
            style={
              {
                "color": "transparent",
              }
            }
            width={20}
          />
        </div>
      </td>
    </tr>
    <tr
      className="[&>td]:pt-10 [&>td]:first:pt-0"
    >
      <td
        className="flex flex-col justify-start break-keep pr-2.5 align-top"
      >
        <h2
          className="my-0 max-w-lg font-bold"
        >
          Brave New World
        </h2>
        <div
          className="text-sm"
        >
          Aldous Huxley
        </div>
      </td>
      <td
        className="align-top text-sm max-smaller:inline"
      >
        Fiction
      </td>
      <td
        className="align-top"
      >
        <div
          className="mx-auto h-5 w-5"
        >
          <img
            alt="Completed Book Icon"
            data-nimg="1"
            decoding="async"
            height={20}
            loading="lazy"
            onError={[Function]}
            onLoad={[Function]}
            sizes="20px"
            src="/_next/image?url=%2Fimg.jpg&w=3840&q=75"
            srcSet="/_next/image?url=%2Fimg.jpg&w=16&q=75 16w, /_next/image?url=%2Fimg.jpg&w=32&q=75 32w, /_next/image?url=%2Fimg.jpg&w=48&q=75 48w, /_next/image?url=%2Fimg.jpg&w=64&q=75 64w, /_next/image?url=%2Fimg.jpg&w=96&q=75 96w, /_next/image?url=%2Fimg.jpg&w=128&q=75 128w, /_next/image?url=%2Fimg.jpg&w=256&q=75 256w, /_next/image?url=%2Fimg.jpg&w=384&q=75 384w, /_next/image?url=%2Fimg.jpg&w=640&q=75 640w, /_next/image?url=%2Fimg.jpg&w=750&q=75 750w, /_next/image?url=%2Fimg.jpg&w=828&q=75 828w, /_next/image?url=%2Fimg.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimg.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimg.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimg.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimg.jpg&w=3840&q=75 3840w"
            style={
              {
                "color": "transparent",
              }
            }
            width={20}
          />
        </div>
      </td>
    </tr>
    <tr
      className="[&>td]:pt-10 [&>td]:first:pt-0"
    >
      <td
        className="flex flex-col justify-start break-keep pr-2.5 align-top"
      >
        <h2
          className="my-0 max-w-lg font-bold"
        >
          Modern Conspiracies in America: Separating Fact from Fiction
        </h2>
        <div
          className="text-sm"
        >
          Michael D. Gambone
        </div>
      </td>
      <td
        className="align-top text-sm max-smaller:inline"
      >
        Nonfiction
      </td>
      <td
        className="align-top"
      >
        <div
          className="mx-auto h-5 w-5"
        >
          <img
            alt="Completed Book Icon"
            data-nimg="1"
            decoding="async"
            height={20}
            loading="lazy"
            onError={[Function]}
            onLoad={[Function]}
            sizes="20px"
            src="/_next/image?url=%2Fimg.jpg&w=3840&q=75"
            srcSet="/_next/image?url=%2Fimg.jpg&w=16&q=75 16w, /_next/image?url=%2Fimg.jpg&w=32&q=75 32w, /_next/image?url=%2Fimg.jpg&w=48&q=75 48w, /_next/image?url=%2Fimg.jpg&w=64&q=75 64w, /_next/image?url=%2Fimg.jpg&w=96&q=75 96w, /_next/image?url=%2Fimg.jpg&w=128&q=75 128w, /_next/image?url=%2Fimg.jpg&w=256&q=75 256w, /_next/image?url=%2Fimg.jpg&w=384&q=75 384w, /_next/image?url=%2Fimg.jpg&w=640&q=75 640w, /_next/image?url=%2Fimg.jpg&w=750&q=75 750w, /_next/image?url=%2Fimg.jpg&w=828&q=75 828w, /_next/image?url=%2Fimg.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimg.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimg.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimg.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimg.jpg&w=3840&q=75 3840w"
            style={
              {
                "color": "transparent",
              }
            }
            width={20}
          />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
